Filter events by labels

dnephin opened the pull request:
Resolves #16458

This is the smallest change I could see to add support for this feature.  I think some refactoring of `getEvents()` would make it easier to unit test, but I wanted to try this first.

Feedback appreciated

dnephin committed:
Resolves #16458 - filter events by labels.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

tiborvass commented on integration-cli/docker_cli_events_test.go:
```diff
@@ -383,6 +383,48 @@ func (s *DockerSuite) TestEventsFilterImageName(c *check.C) {
+	}
+
+	if count1 == 0 {
+		c.Fatalf("Expected events from the labelled container, but got none")
```
`labeled` (same below)

dnephin replied:
I was going with the Canadian/non-US english version http://grammarist.com/spelling/label/, but I can change it :)

tiborvass commented:
I like this, +1

calavera commented:
Moved to code review. @dnephin don't forget that this change is going to require also documentation.

calavera commented on api/server/daemon.go:
```diff
@@ -127,14 +127,36 @@ func (s *Server) getEvents(ctx context.Context, w http.ResponseWriter, r *http.R
 		return c.ID
 	}
 
+	getLabels := func(id string) map[string]string {
```
can we define these functions out of the main handler?

dnephin replied:
sounds good

dnephin commented:
Ok, moved the nested functions out, and added docs.  My docs commit includes some cleanup of end-of-line whitespace (which was done automatically by [my git commit hooks](https://gist.github.com/dnephin/c7776ecb5cdcf02eb1a6)). I hope that's alright.

cpuguy83 commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -1351,12 +1351,12 @@ or being killed.
         the resulting image in case of success.
--   **remote** – A Git repository URI or HTTP/HTTPS URI build source. If the 
-        URI specifies a filename, the file's contents are placed into a file 
+-   **remote** – A Git repository URI or HTTP/HTTPS URI build source. If the
```
This whitespace is intentional, it denotes a newline in markdown.

cpuguy83 commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -1351,12 +1351,12 @@ or being killed.
 Query Parameters:
 
--   **dockerfile** - Path within the build context to the Dockerfile. This is 
+-   **dockerfile** - Path within the build context to the Dockerfile. This is
```
This whitespace is intentional, it denotes a newline in markdown.

thaJeztah replied:
True. I checked here, don't think it's a problem in this case

cpuguy83 commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -1351,12 +1351,12 @@ or being killed.
--   **remote** – A Git repository URI or HTTP/HTTPS URI build source. If the 
-        URI specifies a filename, the file's contents are placed into a file 
+-   **remote** – A Git repository URI or HTTP/HTTPS URI build source. If the
+        URI specifies a filename, the file's contents are placed into a file
```
This whitespace is intentional, it denotes a newline in markdown.

cpuguy83 commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -2506,7 +2507,7 @@ from **200 OK** to **101 UPGRADED** and resends the same headers.
 ## 3.3 CORS Requests
 
-To set cross origin requests to the remote api please give values to 
+To set cross origin requests to the remote api please give values to
```
This whitespace is intentional, it denotes a newline in markdown.

dnephin replied:
I built the docs to investigate this. I think most of these are unnecessary (the list handles things just fine).  This one however I can see being useful to prevent the line wrap of the command below.

I'll revert these changes and disable this pre-commit for markdown files.

dnephin replied:
I don't think the markdown engine (hugo?) does anything with this whitespace, I tried restoring it and it's not forcing a newline here.

cpuguy83 commented:
At least, I'm assuming that whitespace is all intentional! ping @moxiegirl

cpuguy83 commented on integration-cli/docker_cli_events_test.go:
```diff
@@ -383,6 +383,48 @@ func (s *DockerSuite) TestEventsFilterImageName(c *check.C) {
+	if len(events) == 0 {
+		c.Fatalf("Expected events but found none for the label %s", label)
+	}
+	count1 := 0
```
Maybe just use bools for these?

cpuguy83 commented on integration-cli/docker_cli_events_test.go:
```diff
@@ -383,6 +383,48 @@ func (s *DockerSuite) TestEventsFilterImageName(c *check.C) {
+	for _, e := range events {
+		if strings.Contains(e, container1) {
+			count1++
+		} else if strings.Contains(e, container2) {
```
Here we can probably just do `c.Assert(e, check.Not(checker.Contains), container2)`

dnephin replied:
Cool, I was just copying the existing test, but I like this suggestion better, a lot more concise.

cpuguy83 commented on api/server/daemon.go:
```diff
@@ -178,3 +181,23 @@ func (s *Server) getEvents(ctx context.Context, w http.ResponseWriter, r *http.R
 }
+
+func getLabels(d *daemon.Daemon, id string) map[string]string {
+	container, err := d.Get(id)
```
This is going to be a no-go since we are making the effort to move any such interactions into the daemon itself rather than having in it in the api layer.
Maybe we can move the majority of the things in the `getEvents` function into the daemon, and this can go along with it?

dnephin replied:
That makes a lot of sense. I did get the feeling this was the wrong place for it, but didn't want to go moving things around without more context.  

I'll get more of this moved into Daemon.

dnephin commented:
I'm pretty sure the whitespace doesn't do anything, I'll wait for confirmation from @moxiegirl

moxiegirl commented:
checking with a make...

dnephin commented:
@moxiegirl my change isn't related to `cpuset` or deprecating anything, not sure what you mean.

The docs changes were just cleanup done by my automated git hooks to remove end-of-line whitespace.

moxiegirl commented:
@dnephin ugh...I was reading the green lines.  And I'll file an issue for the other stuff because it was missed.

dnephin commented:
@cpuguy83 ok, I've cleaned up the assertions, and moved a lot of the filtering logic out of  the api handler and into the daemon package.  

I'm not sure what you'll think of the `EventFilter` type. The only way I can see around that (and the `GetEventFilter` method on daemon), would be a significant change to the data stored by `Event.Log`. Instead of only storing an ID, it would need to store an object which exposes a way to normalize id/name, and to extract labels (basically store the container or image struct).  I think that would be the way I'd design it.  I'd rather avoid that large change for now, and do it as a follow up (if you think that's the way to go).

dnephin commented:
Ok, I realized there is another issue with the lookup of labels. If the container or image is removed before the API call to events (with a filter on label), and there is a `since=...` param, you actually miss events.  The call to lookup labels fails (because the thing is already removed), so it doesn't match the filter.

This is probably an isolated edge case.  Do you think it necessitates the changes to events I was suggesting above?

dnephin committed:
Documentation for filtering events by label

Signed-off-by: Daniel Nephin <dnephin@docker.com>

cpuguy83 commented on api/server/httputils/form.go:
```diff
@@ -23,7 +23,7 @@ func BoolValueOrDefault(r *http.Request, k string, d bool) bool {
-// Int64ValueOrZero parses a form value into a int64 type.
+// Int64ValueOrZero parses a form value into an int64 type.
 // It returns 0 if the parsing fails.
 func Int64ValueOrZero(r *http.Request, k string) int64 {
```
Can this call `Int64ValueOrDefault(r, k, 0)`?

dnephin replied:
It still has to handle the error, but yes, I think it would be good to have it call the other function for the rest of the logic

cpuguy83 commented on api/server/router/local/info.go:
```diff
@@ -54,26 +52,27 @@ func (s *router) getInfo(ctx context.Context, w http.ResponseWriter, r *http.Req
 	return httputils.WriteJSON(w, http.StatusOK, info)
 }
 
+func buildOutputEncoder(w http.ResponseWriter) *json.Encoder {
```
this could probably go into httputils, it's a common pattern in the API.

dnephin replied:
I grpe'ed for other uses of `outStream.Write(nil)` and only found one that didn't set the encoder or content type.

If you think this will be used in the future I can move it now, but it doesn't appear to be used at this time.

tiborvass commented on daemon/events/filter.go:
```diff
@@ -0,0 +1,64 @@
+)
+
+// EventFilter can filter out docker events from a stream
+type EventFilter struct {
```
it's already in package events, so the name `Filter` should suffice

jfrazelle commented on integration-cli/docker_cli_events_test.go:
```diff
@@ -401,7 +460,7 @@ func (s *DockerSuite) TestEventsFilterContainer(c *check.C) {
 	checkEvents := func(id string, events []string) error {
 		if len(events) != 4 { // create, attach, start, die
-			return fmt.Errorf("expected 3 events, got %v", events)
+			return fmt.Errorf("expected 4 events, got %v", events)
```
why?

jfrazelle replied:
if the test didnt change

dnephin replied:
Isn't the error message wrong here? The check is for 4 events, but error text says it expected 3.

dnephin replied:
Alternatively I could change it to use assert which would handle the messaging without this issue.

jfrazelle replied:
oh lol nice, i didnt notice that

dnephin commented:
Applied `'s/EventFilter/Filter/g'` and changed `Int64ValueOrZero` to use `Int64ValueOrZero

cpuguy83 commented on integration-cli/docker_cli_events_test.go:
```diff
@@ -383,6 +384,64 @@ func (s *DockerSuite) TestEventsFilterImageName(c *check.C) {
+
+	dockerCmd(c, "tag", name, "labelfiltertest:tag1")
+	dockerCmd(c, "tag", name, "labelfiltertest:tag2")
+
```
Looks like we need a case for an event that would actually be filtered out.

dnephin replied:
I believe that is covered by `TestEventsFilterLabels` I create two containers, but assert that I only get events for one.

In the case of images the filtering is applied using the same code as container filtering, so that should be covered already?

I will add tag "busybox" here to show it working for images as well

dnephin replied:
extra tag added

cpuguy83 commented:
I don't like passing in `daemon.GetLabels`, but I can't think of a better way at the moment.
So, LGTM.

jfrazelle commented:
LGTM

jfrazelle commented:
can you squash the commits please

vdemeester commented:
Once squashed, LGTM ;-)

dnephin commented:
squashed to two commits (one for code one for docs)

cpuguy83 commented:
Moving to docs review

thaJeztah commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -2041,6 +2041,7 @@ Query Parameters:
   -   `event=<string>`; -- event to filter
   -   `image=<string>`; -- image to filter
   -   `container=<string>`; -- container to filter
+  -   `label=<string>`; -- image and container label to filter
```
not your fault, but now that this list is getting longer, would you mind putting the list in alphabetical order?

thaJeztah commented on docs/reference/commandline/events.md:
```diff
@@ -51,6 +51,7 @@ The currently supported filters are:
 * container
 * event
 * image
+* labels
```
should be "label" (singular)

thaJeztah replied:
Hm, I think this is the first filter we have that has a *key* and *value*, or is that not supported? Can I, for example, do:
`--filter label=foo` to show al events coming from a container/image with a label `foo`, irregardless of its value?
Can I also filter on the *value* of that label? `--filter label=foo=bar`

Because this is new here, I think we need a small example, and some change in the wording above

dnephin replied:
`key=value` is supported, it works the same way as filters on images or containers

thaJeztah replied:
but we don't currently have a filter, that uses `--filter=<key>=<name>=<value>`, at least, no example, because with labels the user is able to set both the *name* and the *value* of the label.

thaJeztah replied:
where `<key>` in the above, is the name of the filter (`"label"`)

dnephin replied:
True, I'll copy an example from the images.md docs

dnephin commented:
Fixed the docs mistake (and sorted filter list)

cpuguy83 commented on daemon/daemon.go:
```diff
@@ -528,6 +529,36 @@ func (daemon *Daemon) GetByName(name string) (*Container, error) {
 }
 
+// GetFilter returns a filters.Filter for a set of filters
+func (daemon *Daemon) GetFilter(filter filters.Args) *events.Filter {
```
You know what... sorry... just noticed this function name is not good since it is specifically for events.

dnephin replied:
Ah right, I renamed this with the `s/EventFilter/Filter/`, but I should change this back.

thaJeztah commented:
@dnephin oops, just added one more comment inline: https://github.com/docker/docker/pull/16530#discussion_r41064013.

Also, I think we need a mention in https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v121-api-changes

Hm, looks like the man page doesn't actually mention which filters are available; https://github.com/docker/docker/blob/master/man/docker-events.1.md. Would be nice to add it there, but since it's not there currently, I'll leave that up to you :+1:

dnephin commented:
Fixed `s/GetFilter/GetEventFilter`, and added to the api changes doc

dnephin commented:
Updated filter examples in command line reference for events

thaJeztah commented:
Thanks! docs LGTM

moxiegirl commented:
LGTM

runcom commented:
Thanks @dnephin!
